setversion.py

```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python
+
+# CbAPI Project Version Number Setting Script
+# AGRB 11/7/2019
+
+import sys
+import os
+import re
+import argparse
+from datetime import date
+
+
+def readme_rewriter(line, ctxt):
+    expr = ctxt.get("readme_expr", None)
+    if not expr:
+        expr = re.compile(r"^\*\*Latest Version:")
+        ctxt["readme_expr"] = expr
+    if expr.match(line):
+        return "**Latest Version: {0}**\n".format(ctxt["version"])
+    return None
+
+
+def changelog_rewriter(line, ctxt):
+    expr = ctxt.get("changelog_expr", None)
+    if not expr:
+        expr = re.compile(r"^\.\. top-of-changelog")
+        ctxt["changelog_expr"] = expr
+    if expr.match(line):
+        datestr = date.today().strftime("%B %d, %Y")
+        cl1 = "CbAPI {0} - Released {1}".format(ctxt["version"], datestr)
+        cl2 = "".ljust(len(cl1), "-")
+        updates = "\n\nUpdates\n\n.. (add your updates here)\n"
+        return line + "\n" + cl1 + "\n" + cl2 + updates
+    return None
+
+
+def doc_conf_rewriter(line, ctxt):
+    vexpr = ctxt.get("doc_version_expr", None)
```

Comment:
I see these ctxt.get in a few functions, but I don't see where those are being exposed

Reply:
ah I see you did this to not have to compile the regular expression each time

Reply:
Those are essentially using elements of the context as a cache, so the regular expression doesn't have to be recompiled for each line.